### ISSUE TITLE
writef leading to linking error when making binary (some platforms)

### ISSUE THREAD

**Issue by kwaters4** [chapel-lang/chapel#21576]
### Summary of Problem

I was working on one of my machines (RockyLinux) and using a `writef` statement  and generates a linker error making the binary. I was unable to reproduce this on my MacOS (BigSur) machine. 

I have been building Chapel following the [quick start instructions](https://chapel-lang.org/docs/usingchapel/QUICKSTART.html). In terms of Chapel itself the MacOS and RockyLinux builds are identical outside of the tools used to build it. I have provided the configurations of the install with the problem below. 

Please let me know what additional information is needed.

### Steps to Reproduce

**Source Code:**

```chapel
writef("My favorite %s is %i\n", "number", 7);
```

**ATO Link:**
I think I was able to reproduce the error using [ATO](https://ato.pxeger.com/run?1=m70sOSOxIDVnwYKlpSVpuhY39cqLMktS0zSUfCsV0hLL8kE8BdVihcxiBdXMmDwlHQWlvNLcpNQiIMtc0xqiC6oZZggA). I am not 100%, the output is as follows:

```
/ATO/runner: line 6:     3 Killed                  /ATO/yargs % /ATO/options chpl % /ATO/code.chpl -o /ATO/exe
yargs: execvp: No such file or directory
```

I am not sure if ATO would print out linker errors.

**Compile command:**
chpl foo.chpl

### Error 
(Some sanitation occured to clean up the path and spacing)
```
/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(re2.o): In function `re2::Fanout(re2::Prog*, std::vector<int, std::allocator<int> >*)':
re2.cc:(.text+0x13fa): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(re2.o): In function `LogMessage::LogMessage(char const*, int)':
re2.cc:(.text._ZN10LogMessageC2EPKci[_ZN10LogMessageC2EPKci]+0x18): undefined reference to `std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >::basic_ostringstream()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(regexp.o): In function `re2::Regexp::ConcatOrAlternate(re2::RegexpOp, re2::Regexp**, int, re2::Regexp::ParseFlags, bool)':
regexp.cc:(.text+0xe51): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(regexp.o): In function `re2::ConvertRunesToBytes(bool, int*, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
regexp.cc:(.text+0x2057): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::reserve()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(regexp.o): In function `std::deque<re2::WalkState<int>, std::allocator<re2::WalkState<int> > >::_M_reallocate_map(unsigned long, bool)':
regexp.cc:(.text._ZNSt5dequeIN3re29WalkStateIiEESaIS2_EE17_M_reallocate_mapEmb[_ZNSt5dequeIN3re29WalkStateIiEESaIS2_EE17_M_reallocate_mapEmb]+0x14d): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(regexp.o): In function `std::deque<re2::WalkState<re2::Range>, std::allocator<re2::WalkState<re2::Range> > >::_M_reallocate_map(unsigned long, bool)':
regexp.cc:
(.text._ZNSt5dequeIN3re29WalkStateINS0_5RangeEEESaIS3_EE17_M_reallocate_mapEmb[_ZNSt5dequeIN3re29WalkStateINS0_5RangeEEESaIS3_EE17_M_reallocate_mapEmb]+0x14d): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(bitstate.o): In function `re2::BitState::GrowStack()':
bitstate.cc:(.text+0x98): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(bitstate.o): In function `re2::BitState::Push(int, char const*)':
bitstate.cc:(.text+0x29b): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(bitstate.o): In function 
`re2::BitState::Search(re2::StringPiece const&, re2::StringPiece const&, bool, bool, re2::StringPiece*, int)':
bitstate.cc:(.text+0xa1c): undefined reference to `std::__throw_bad_array_new_length()'

/Chapel/source/chapel-1.29.0/third-party/re2/install/linux64-x86_64-native-llvm-none/lib/libre2.a(compile.o):compile.cc:(.text+0x25d): 
more undefined references to `std::__throw_bad_array_new_length()' follow

clang-14: error: linker command failed with exit code 1 (use -v to see invocation)

error: Make Binary - Linking
```

### Configuration Information

- Output of `chpl --version`:
```
 chpl version 1.29.0
  built with LLVM version 14.0.0
Copyright 2020-2022 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)
```
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native *
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: bundled
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
```
gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-15)

clang version 14.0.6 (Red Hat 14.0.6-1.module+el8.7.0+1080+d88dc670)
Target: x86_64-redhat-linux-gnu
```

**Comment by cassella** [comment]
@kwaters4 It works for me (on linux).  The errors look like some sort of incompatibility between the Chapel's re2 library and some system c++ library.

Does it help if you build with `CHPL_RE2=none chpl foo.chpl` ?

Did you build that Chapel release from source?  If so, guessing that a system update has changed something, I'd try rebuilding it (after a full `make clobber`).

I get the same ATO error with `writeln("hello world")`, so I think something's just busted over there.  :(

**Comment by kwaters4** [comment]
@cassella Thanks for the ideas and the quick response! I just went straight to the rebuild and it is now working. 

I did not think of trying something else with ATO, that was just a red herring then. ATO is still presenting with issues now.

**Comment by bradcray** [comment]
w.r.t. the ATO issue, I've been in mourning the past week because I'd come to rely on it so much in my day-to-day work and it hasn't been working for anything since about the 5th.  I've asked the administrator about it and he [attributed it to something that has since been resolved](https://chat.stackexchange.com/transcript/message/62932305#62932305).  I've opened https://github.com/chapel-lang/chapel/issues/21588 to at least acknowledge that we're aware it's not working.
